kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,624 @@
+                (STRING, column_name!("remove.path")),
+                (STRING, column_name!("remove.deletionVector.storageType")),
+                (STRING, column_name!("remove.deletionVector.pathOrInlineDv")),
+                (INTEGER, column_name!("remove.deletionVector.offset")),
+                (STRING, column_name!("cdc.path")),
+                (LONG, column_name!("commitInfo.timestamp")),
+                (STRING, column_name!("metaData.schemaString")),
+                (INTEGER, column_name!("protocol.minReaderVersion")),
+                (INTEGER, column_name!("protocol.minWriterVersion")),
+                (string_list.clone(), column_name!("protocol.readerFeatures")),
+                (string_list, column_name!("protocol.writerFeatures")),
+            ];
+            let (types, names) = types_and_names.into_iter().unzip();
+            (names, types).into()
+        });
+        NAMES_AND_TYPES.as_ref()
+    }
+
+    fn visit<'b>(
+        &mut self,
+        row_count: usize,
+        getters: &[&'b dyn crate::engine_data::GetData<'b>],
+    ) -> DeltaResult<()> {
+        let expected_getters = 12;
+        require!(
+            getters.len() == expected_getters,
+            Error::InternalError(format!(
+                "Wrong number of Phase1Visitor getters: {}",
+                getters.len()
+            ))
+        );
+
+        for i in 0..row_count {
+            if !self.selection_vector[i] {
+                println!("skipping");
+                continue;
+            }
+            if let Some(path) = getters[0].get_str(i, "add.path")? {
+                println!("Found add path");
+                self.scanner.add_paths.insert(path.to_string());
+            } else if let Some(path) = getters[1].get_str(i, "remove.path")? {
+                println!("Found remove path");
+                let deletion_vector = visit_deletion_vector_at(i, &getters[2..=4])?;
+                self.scanner
+                    .remove_dvs
+                    .insert(path.to_string(), DvInfo { deletion_vector });
+            } else if getters[5].get_str(i, "cdc.path")?.is_some() {
+                println!("Found cdc path");
+                self.scanner.has_cdc_action = true;
+            } else if let Some(timestamp) = getters[6].get_long(i, "commitInfo.timestamp")? {
+                println!("Found timestamp");
+                self.scanner.timestamp = timestamp;
+            } else if let Some(schema) = getters[7].get_str(i, "metaData.schemaString")? {
+                println!("Found metadata");
+                // TODO: Validate that the schema is as expected
+            } else if let Some(min_reader_version) =
+                getters[8].get_int(i, "protocol.min_reader_version")?
+            {
+                println!("Found protocol");
+                let protocol =
+                    ProtocolVisitor::visit_protocol(i, min_reader_version, &getters[8..=11])?;
+                protocol.ensure_read_supported()?;
```

Comment:
I think the error will be confusing if the _visit_ fails due to a protocol that isn't supported. Could we just store the protocol in the scanner and then `ensure` in `phase1` call above?